### PR TITLE
Support UUST2

### PR DESCRIPTION
**Pull request by at-wat**
The new revision of UST (UUST2) doesn't handle the TM command immediately, which doesn't fit the SCIP2 protocol's design.
This PR adds a new logic to estimate the precise timestamp of UUST2; however, it takes longer to stop scanning data for synchronization.

### CODE REVIEW

File: src/clock_estimator_uust1.cpp [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+  if (n < MIN_SYNC_SAMPLES)
+  {
+    return false;
+  }
+  if (n >= MAX_SYNC_SAMPLES)
+  {
+    return true;
+  }
+  const OriginFracPart overflow_range = originFracOverflow();
+  return overflow_range.t_max_ > overflow_range.t_min_;
```
Comment (f-fl0) [inline src/clock_estimator_uust1.cpp:70]: Should we check if `overflow_range` is valid i.e. `return overflow_range && overflow_range.t_max_ > overflow_range.t_min_;`?

File: include/urg_stamped/device_state_estimator.h [hunk 2]
```diff
@@ -212,52 +247,132 @@ class Estimator
...
+  int cnt_samples_;
 };
 
-class EstimatorUTM : public Estimator
+class ScanEstimatorUTM : public ScanEstimator
 {
 public:
-  inline explicit EstimatorUTM(const ros::Duration& ideal_scan_interval)
-    : Estimator(ideal_scan_interval)
+  inline explicit ScanEstimatorUTM(const ClockEstimator::Ptr clock_estim, const ros::Duration& ideal_scan_interval)
```
Comment (f-fl0) [inline include/urg_stamped/device_state_estimator.h:374]: Same here about `explicit`.

File: include/urg_stamped/device_state_estimator.h [hunk 3]
```diff
@@ -276,14 +391,14 @@ class EstimatorUTM : public Estimator
...
+  FRIEND_TEST(ScanEstimatorUTM, PushScanSampleRaw);
 };
 
-class EstimatorUST : public Estimator
+class ScanEstimatorUST : public ScanEstimator
 {
 public:
-  inline explicit EstimatorUST(const ros::Duration& ideal_scan_interval)
-    : Estimator(ideal_scan_interval)
+  inline explicit ScanEstimatorUST(const ClockEstimator::Ptr clock_estim, const ros::Duration& ideal_scan_interval)
```
Comment (f-fl0) [inline include/urg_stamped/device_state_estimator.h:400]: Same here about `explicit`.

**Review by f-fl0 — APPROVED** [review]
LGTM. I just left some minor comments.

**Review by Taka-Kazu — APPROVED** [review]

File: src/clock_estimator_uust1.cpp (hunk 1 above)
Comment (at-wat) [reply src/clock_estimator_uust1.cpp:70]: As this seems be a bug in the master branch, I'll fix it in a new PR

File: include/urg_stamped/device_state_estimator.h [hunk 4]
```diff
@@ -203,6 +216,28 @@ class Estimator
...
+
+  bool pushClockSample(const ClockSample& clock);
+};
+
+class ScanEstimator
+{
+public:
+  using Ptr = std::shared_ptr<ScanEstimator>;
+
+  inline explicit ScanEstimator(const ClockEstimator::Ptr clock_estim, const ros::Duration& ideal_scan_interval)
```
[the remark this answers is not in the document]
Comment (at-wat) [reply include/urg_stamped/device_state_estimator.h]: Forgot to remove when adding an argument

File: src/clock_estimator_uust1.cpp (hunk 1 above)
Comment (at-wat) [reply src/clock_estimator_uust1.cpp:70]: `overflow_range` is an instance and `t_max_ = t_min_ = 0` when invalid. So, the result is correct even if `originFracOverflow()` failed

**Pull request merged**